feat: turn off notification via email

joobisb opened the pull request:
This PR addresses #14389 

As seen in the attached images, users will receive an option saying `Stop receiving emails like this`. On clicking, the user will be redirected to notification settings and the template of that particular notification type will be disabled.



![image](https://github.com/user-attachments/assets/68952d92-dc67-48d2-a7c3-f33dc45fc336)
![image](https://github.com/user-attachments/assets/3618d2c0-92ff-41ae-9d6f-d83c6db3f4ea)

joobisb committed:
feat: turn off notification via email

joobisb commented:
I have read the CLA Document and I hereby sign the CLA

joobisb committed:
remove unused imports

joobisb commented:
@dannykopping @stirby could you please have a look at this PR

dannykopping commented on coderd/notifications/enqueuer.go:
```diff
@@ -123,7 +123,8 @@ func (s *StoreEnqueuer) buildPayload(metadata database.FetchNewMessageMetadataRo
 
-		NotificationName: metadata.NotificationName,
+		NotificationName:       metadata.NotificationName,
+		NotificationTemplateID: metadata.NotificationTemplateID.String(),
```
With the addition, we should bump the payload version by a minor release since it's purely additive.

dannykopping commented on coderd/notifications/types/payload.go:
```diff
@@ -7,13 +7,12 @@ package types
-	Labels  map[string]string `json:"labels"`
+	NotificationName       string            `json:"notification_name"`
+	NotificationTemplateID string            `json:"notification_template_id"`
+	UserID                 string            `json:"user_id"`
```
Let's keep the contextual groupings please.

dannykopping commented on site/src/pages/UserSettingsPage/NotificationsPage/NotificationsPage.tsx:
```diff
@@ -60,6 +62,27 @@ export const NotificationsPage: FC = () => {
 	const updatePreferences = useMutation(
 		updateUserNotificationPreferences(user.id, queryClient),
 	);
+	const [searchParams] = useSearchParams();
```
@BrunoQuaresma could you please review this?

BrunoQuaresma replied:
Sure thing!

dannykopping commented on coderd/notifications/dispatch/smtp/html.gotmpl:
```diff
@@ -26,6 +26,7 @@
       <div style="border-top: 1px solid #e2e8f0; color: #475569; font-size: 12px; margin-top: 64px; padding-top: 24px; line-height: 1.6;">
         <p>&copy;&nbsp;{{ current_year }}&nbsp;Coder. All rights reserved&nbsp;-&nbsp;<a href="{{ base_url }}" style="color: #2563eb; text-decoration: none;">{{ base_url }}</a></p>
         <p><a href="{{ base_url }}/settings/notifications" style="color: #2563eb; text-decoration: none;">Click here to manage your notification settings</a></p>
+        <p><a href="{{ base_url }}/settings/notifications?unsubscribe={{ .NotificationTemplateID }}" style="color: #2563eb; text-decoration: none;">Stop receiving emails like this</a></p>
```
I think we should try keep terminology consistent here please. We don't refer to "subscribing" to notifications, but rather than templates are "enabled" or "disabled" for a user.

dannykopping commented on site/src/pages/UserSettingsPage/NotificationsPage/NotificationsPage.tsx:
```diff
@@ -60,6 +62,27 @@ export const NotificationsPage: FC = () => {
+	}, [unsubscribeTemplateId]);
+
+	const handleUnsubscribe = async (templateId: string) => {
+		await updatePreferences.mutateAsync({
```
Error handling?

dannykopping reviewed (commented):
This is great work @joobisb 👍 

I think we should probably add a frontend test as well, and show a message to the user which template was disabled by highlighting it.

A few notes but nothing major.
Thank you for your contribution! 🎉

joobisb committed:
addressed the comments

joobisb commented:
> This is great work @joobisb 👍
> 
> I think we should probably add a frontend test as well, and show a message to the user which template was disabled by highlighting it.
> 
> A few notes but nothing major. Thank you for your contribution! 🎉

@dannykopping I could not see tests related to the `NotificationsPage`. If there are any, could you point me to where to add them?

Apart from that, the rest of the comments are addressed, please have a look.

dannykopping commented on coderd/notifications/Archive.zip:
Please remove this.

dannykopping commented on site/src/pages/UserSettingsPage/NotificationsPage/NotificationsPage.tsx:
```diff
@@ -60,6 +62,41 @@ export const NotificationsPage: FC = () => {
+			);
+		} catch (error) {
+			console.error(error);
+			displayError("Error on disabling notification");
```
```suggestion
			displayError("An error occurred when attempting to disable the requested notification");
```

dannykopping replied:
Is this only catching the `throw` on line 86?
We should also log the error response (if present) of the `mutateAsync` call.

joobisb replied:
this would catch the `mutateAsync` errors as well

dannykopping commented:
> @dannykopping I could not see tests related to the `NotificationsPage`. If there are any, could you point me to where to add them?

@BrunoQuaresma can assist with that, I'm not too familiar with our frontend I'm afraid.

BrunoQuaresma commented on site/src/pages/UserSettingsPage/NotificationsPage/NotificationsPage.tsx:
```diff
@@ -60,6 +62,41 @@ export const NotificationsPage: FC = () => {
+		}
+	}, [templateId, templatesByGroup.isSuccess, templatesByGroup.data]);
+
+	const disableTemplate = async (templateId: string) => {
+		try {
+			await updatePreferences.mutateAsync({
+				template_disabled_map: {
+					[templateId]: true,
+				},
+			});
+
+			const allTemplates = Object.values(templatesByGroup.data ?? {}).flat();
+			const template = allTemplates.find((t) => t.id === templateId);
+
+			if (!template) {
+				throw new Error(`Template with ID ${templateId} not found`);
+			}
+
+			displaySuccess(`${template.name} notification has been disabled`);
+
+			queryClient.invalidateQueries(
+				userNotificationPreferences(user.id).queryKey,
+			);
+		} catch (error) {
+			console.error(error);
+			displayError("Error on disabling notification");
```
I feel the `disableTemplate` could be a mutation inside of the `queries/notifications`. Eg.:

```tsx
const queryClient = useQueryClient()
const disableMutation = useMutation(disableNotification(userId, queryClient))

useEffect(() => {
  try {
    disableMutation
      .mutateAsync()
      .then(() => {
        displaySuccess("...")
      }).catch(() => {
        displayError("...")
      })
  }
}, [...])
```

BrunoQuaresma commented on site/src/pages/UserSettingsPage/NotificationsPage/NotificationsPage.tsx:
I need to remember to check for the following scenarios:

1. When a valid template ID is disabled and the request is successful, check if the success message is displayed. Also, ensure that the UI is updated correctly. Verify if the request is being sent properly. Additionally, confirm if the error message is displayed when the request fails.

2. When a not found template ID is disabled, check if an error message is displayed.

You can check how we do that using [Storybook interaction tests](https://storybook.js.org/tutorials/ui-testing-handbook/react/en/interaction-testing/) on the [`NotificationsPage.stories.tsx`](https://github.com/coder/coder/blob/95a7c0c4f087744a22c2e88dd3c5d30024d5fb02/site/src/pages/DeploySettingsPage/NotificationsPage/NotificationsPage.stories.tsx).

joobisb replied:
I have verified these scnearios

dannykopping replied:
@joobisb can you add a test please? Manual verification is good but it only helps us know if this works currently, and doesn't catch future degradations.

BrunoQuaresma replied:
@joobisb we are close, we just need to automate these tests using the way I shared before. Thanks for your hard work! 🙏

joobisb replied:
@BrunoQuaresma the tests needed to be added to `NotificationsPage.stories.tsx` right ?

joobisb replied:
> @joobisb we are close, we just need to automate these tests using the way I shared before. Thanks for your hard work! 🙏

done, please have a look

BrunoQuaresma reviewed (commented):
@joobisb good work! I just left a few comments but let me know if you need any extra explanation or help. Thank you!

stirby commented:
@joobisb, thanks for submitting this! The copy looks fine to me. Leaving approval to Danny and Bruno.

joobisb committed:
chore: refactored notifications page

joobisb committed:
refactored and added tests

joobisb commented:
> @joobisb good work! I just left a few comments but let me know if you need any extra explanation or help. Thank you!

@BrunoQuaresma 

I have addressed the comments. Could you please take a look?

BrunoQuaresma commented:
@joobisb great! I will check the chromatic result in a few minutes when it is ready. If everything is okay, I will approve it. ✅

joobisb commented:
> @joobisb, it seems like the tests you wrote in the storybook are failing. Have you had a chance to check them locally?

@BrunoQuaresma 
Yeah, it's failing mostly due to issues with testing `useEffect`. I tried a few things, but they didn't work. I'm not that familiar with the front end. Do you have any test cases that I can refer to for such scenarios?

BrunoQuaresma commented:
@joobisb I think we solved some of those errors last week. Could you please update this branch with the main please? 🙏

BrunoQuaresma commented:
I will try to make some time today to look at the code and see how I can help.

joobisb committed:
Merge remote-tracking branch 'origin/main' into email-notification-disable

joobisb commented:
> I will try to make some time today to look at the code and see how I can help.

@BrunoQuaresma I've updated the branch with main

BrunoQuaresma committed:
Simplify effect

BrunoQuaresma committed:
Fix stories

BrunoQuaresma committed:
Fix fmt
